src/ripple/app/ledger/impl/LedgerCleaner.cpp

```diff
@@ -96,20 +101,24 @@ class LedgerCleanerImp
     //
     //--------------------------------------------------------------------------
 
-    void onPrepare ()
+    void onPrepare () override
     {
     }
 
-    void onStart ()
+    void onStart () override
     {
-        startThread();
+        thread_ = std::thread {&LedgerCleanerImp::run, this};
     }
 
-    void onStop ()
+    void onStop () override
     {
-        m_journal.info << "Stopping";
-        signalThreadShouldExit();
-        notify();
+        JLOG (j_.info) << "Stopping";
+        {
+            std::lock_guard<std::mutex> lock (mutex_);
+            state_ = State::shouldExit;
+            wakeup_.notify_one();
```

Comment:
Hmmm.  Interesting.  Mr. Williams consistently puts the `notify_one()` call under the lock.  But I haven't seen any text where he justifies that.  And Mr. Josuttis puts his calls to `notify_one()` outside the lock.  I agree with your reasoning about the possible performance benefits.  So I'll move the `notify_one()` calls outside the lock.  Thanks.